### PR TITLE
feature: backend sftp

### PR DESCRIPTION
**Pull request by alexisvisco**
Adding the sftp as backend for the cache storage.
Relation with issue #60 

todo:
- [x] init sft client and add it to the backend switch
- [x] implementing Put
- [x] implementing Get
- [x] test, re-test and test and ... test.

### CODE REVIEW

**Comment by alexisvisco** [comment]
To test the sftp storage backend: 

Run `docker run -p 22:22 -d atmoz/sftp foo:pass:::upload` this will create an sftp server with an user foo and he can only access to `/home/foo/upload`

Then replace TestSFTPTruth with this code: 

```go
func TestSFTPTruth(t *testing.T) {
	cli, err := InitializeSFTPBackend(SFTPConfig{
		CacheRoot: "/upload",
		Username:  "foo",
		Auth: SSHAuth{
			Password:      "pass",
			Method:        SSHAuthMethodPassword,
		},
		Host:      "localhost",
		Port:      "22",
	}, true)
	if err != nil {
		t.Fatal(err)
	}

	content := "Hello world4"

	// PUT TEST
	file, _ := os.Create("test")
	_, _ = file.Write([]byte(content))
	_, _ = file.Seek(0, 0)
	err = cli.Put("test3.t", file)
	if err != nil {
		t.Fatal(err)
	}
	_ = file.Close()

	// GET TEST
	readCloser, err := cli.Get("test3.t")
	if err != nil {
		t.Fatal(err)
	}
	b, _ := ioutil.ReadAll(readCloser)
	if !bytes.Equal(b, []byte(content))  {
		t.Fatal(string(b), "!=", content)
	}

	_ = os.Remove("test")
}
```

**Comment by alexisvisco** [comment]
Missing test with a ftp server (not a sftp)

**Comment by kakkoyun** [comment]
@alexisvisco Good work so far, thanks a lot. Ping me when you're done.

**Comment by alexisvisco** [comment]
@kakkoyun  Hey, I think I should do another pull request for the FTP part because the library used does not provide any ftp support, SFTP is a protocol to send and retrieve files over ssh.

Maybe you should look for an FTP library and tell me what you want. 
In the meantime I correct the codebeat errors

**Comment by alexisvisco** [comment]
You can merge

**Comment by kakkoyun** [comment]
@alexisvisco Thanks a lot for the work. I will thoroughly review it, please be patient, it won't take long. In the meantime, feel free to create another PR for FTP

File: cache/backend/sftp_test.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package backend
+
+import (
+	"testing"
+)
+
+func TestSFTPTruth(t *testing.T) {
+	t.Skip("skipping backend package tests")
```
Comment (kakkoyun) [inline cache/backend/sftp_test.go]: In order to accept the PR we need some tests. You should actually add the test that you have mentioned in  [your comment](https://github.com/meltwater/drone-cache/pull/68#issuecomment-539940243) in here or you can add your tests what we already have in [plugin tests](https://github.com/meltwater/drone-cache/blob/master/plugin/plugin_test.go). 

I have a separate issue to address issues with testing like separating them into files by their concerned backend or functionality. But for now, we can put them together.

You also mentioned an additional docker container that you need for your tests, it's perfectly fine just add that image in our [`docker-compose file`](https://github.com/meltwater/drone-cache/blob/master/docker-compose.yml) and use [test script](https://github.com/meltwater/drone-cache/blob/master/test) to run your tests.

File: plugin/plugin.go [hunk 2]
```diff
@@ -32,6 +32,7 @@ type (
 
 		S3         backend.S3Config
 		FileSystem backend.FileSystemConfig
+		SFTP       backend.SFTPConfig
```
[the remark this answers is not in the document]
Comment (kakkoyun) [reply plugin/plugin.go]: Also, we have [another file](https://github.com/meltwater/drone-cache/blob/master/DOCS.md) that need to be updated. It's the file we put in [drone-plugin-index](https://github.com/drone/drone-plugin-index)

File: cache/backend/backend.go [hunk 3]
```diff
@@ -88,3 +92,90 @@ func InitializeFileSystemBackend(c FileSystemConfig, debug bool) (cache.Backend,
...
+	Username  string
+	Host      string
+	Port      string
+	Auth      SSHAuth
+}
+
+func InitializeSFTPBackend(c SFTPConfig, debug bool) (cache.Backend, error) {
+	sshClient, err := getSSHClient(c)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (alexisvisco) [reply cache/backend/backend.go]: getSSHClient(c) is now already wrapped in the function getSSHClient

**Comment by kakkoyun** [comment]
@alexisvisco Have you made any progress? Do you need help? Just ping me if you need anything.

**Comment by alexisvisco** [comment]
Hi I have done the test and add a docker-compose service. Unfortunately the continuous integration test does not work and I don't really have time to dive into your drone configuration.

Could you please test manually the sftp config with a drone test ? I have no more time to doing this PR and it's almost finished, just test the config and add details in the https://github.com/meltwater/drone-cache/blob/master/DOCS.md file.

Sorry for the inconvenience.

**Comment by kakkoyun** [comment]
@alexisvisco Thank you very much @alexisvisco, it's all good. I can take it from here. Good work.

To be able to push to this branch directly, could you give me access on your forked repo?

**Comment by alexisvisco** [comment]
@kakkoyun You didn't have it already since it is a fork ? You should have access with this link https://github.com/alexisvisco/drone-cache/invitations

**Pull request merged**